File: 2024_01_03_rust/code/race_condition.cpp
#include <iostream>
#include <thread>

int sharedVariable = 0;

void increment() {
    for (int i = 0; i < 10000; ++i) {
        ++sharedVariable;
    }
}

int main() {
    std::thread thread1(increment);
    std::thread thread2(increment);

    thread1.join();
    thread2.join();

    std::cout << "Expected: 20000, Actual: " << sharedVariable << std::endl;

    return 0;
}

File: 2024_01_03_rust/code/rust/examples/race_condition.rs
use std::thread::spawn;

static mut SHARED_VARIABLE: u32 = 0;

fn increment() {
    for _ in 0..10000 {
        unsafe { SHARED_VARIABLE += 1 };
    }
}

fn main() {
    let thread1 = spawn(increment);
    let thread2 = spawn(increment);

    thread1.join().unwrap();
    thread2.join().unwrap();

    println!("Expected: 20000, Actual: {}", unsafe { SHARED_VARIABLE });
}
